framework/finalizer.go

```diff
@@ -0,0 +1,99 @@
+package framework
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	finalizerName = "operatorkit.giantswarm.io"
+)
+
+type patchSpec struct {
+	Op    string   `json:"op"`
+	Path  string   `json:"path"`
+	Value []string `json:"value"`
+}
+
+func (f *Framework) addFinalizer(obj interface{}) (bool, error) {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return false, err
+	}
+	if accessor.GetDeletionTimestamp() != nil {
+		return true, nil // object has been marked for deletion, we should ignore it.
```

Comment:
I believe that is the right move here. Docs state that this is only added once the object was deleted, afterwards finalizers can only be removed. We should not do anything else with the object other than `DeleteFunc`.
https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/#finalizers

Reply:
Does this happen at all? We only add finalizers on create/update. When the deletion timestamp is not nil we process delete events. So I feel like this code branch is never executed. It would be good to have integration tests for things like this.

Reply:
I added this for the potential problem, that removing a finalizer could be an update event on the object. Checking the `DeletionTimestamp` here lets us ignore that update.

I have not tested yet if that is indeed the behaviour.